[#894] tutorial tester

lavocatt opened the pull request:
This PR brings in a program to test the tutorials under `docs/tutorials`.

Usage (to be executed at the root of the operator)

```
go run test/utils/tutorials/tester.go --help
Usage of /tmp/go-build46072326/b001/exe/tester:
  -dry-run
    	just list what would be executed without doing it
  -ignore-breakpoints
    	ignore the breakpoints
  -interactive
    	prompt to press enter between each chunk
  -list
    	just list the tutorials found
  -no-styling
    	disable spiners in cli
  -quiet
    	disable output
  -start-from string
    	start from a specific stage name
  -timeout int
    	the timeout in minutes for every executed command (default 10)
  -tutorial string
    	Run only a specific tutorial
  -tutorials-root string
    	where to find the tutorials (default "./docs/tutorials")
  -update-tutorials
    	update the chunk output section in the tutorials
  -verbose
    	print more logs
```

## Examples of outputs

### Executing when everything goes well should give this output:

`go run test/utils/tutorials/tester.go --tutorial send_receive_ingress_pem.md`
![output asciicast](https://github.com/user-attachments/assets/ee2222d9-59d2-4e6f-867f-6df2d1eb8567)

### Wrong json metadata

#### bad syntax on the json itself

![output asciicast](https://github.com/user-attachments/assets/eddfd2ef-8ed5-4687-8d30-dcc42e578a82)

#### content is failing static validation

![output asciicast](https://github.com/user-attachments/assets/edc380e5-01e3-45a2-838d-d98bdaef390b)

#### content is failing dynamic validation

![output asciicast](https://github.com/user-attachments/assets/36248e12-ecf7-4bb7-9cd1-8adc6abc59ac)


### When a command fails (for instance by misspelling a command):

![output asciicast](https://github.com/user-attachments/assets/d7634c9c-b8e2-4b1e-b859-c085208eb34f)

#### Verbose output

![output asciicast](https://github.com/user-attachments/assets/be299446-2ef2-4156-9fe6-0138c207b88c)


In this example, the teardown stage for deleting the minikube instance gets executed to cleanup afterwards.

## Interactive mode and breakpoints

In interactive mode you'll get prompted for every commands. Answers can be:
* yes:execute
* no: don't
* all: opt out of interactive mode (until the next breakpoint)
* cancel: execute nothing until the teardown stage.

There's two ways to enter interactive mode, either pass `--intreactive` on the command line. Or add a breakpoint in the chunk you want to stop into.

![output asciicast](https://github.com/user-attachments/assets/26f4e55c-c9e8-455c-a3af-369da46252cf)

## Updating the chunk outputs in the tutorials

Use the `--update-tutorials` option to include the output of the commands inside the original documents.
It's an easy way to keep the document in sync with the commands that are actually executed

## More details

There's a tutorial about how to create a tutorial. Check it out to get the details of the capabilities.

gtully commented on docs/tutorials/send_receive_ingress_pem.md:
````diff
@@ -51,24 +51,29 @@ If you are running your OpenShift cluster locally, you might not be able to
+
+```{"stage":"init"}
+$ kubectl wait pod --all --for=condition=Ready --namespace=send-receive-project
+pod/activemq-artemis-controller-manager-7766ffd9f-vxldz condition met
````
or as example output that is ignored, that is from the wait command, not to be executed. not having a $

lavocatt replied:
Yes that's a sample output. So basically, every command needs to start with a `$` sign, unless we're in a multi-line EOF situation.

gtully commented on docs/tutorials/send_receive_ingress_pem.md:
````diff
@@ -89,7 +98,7 @@ cert-manager-webhook-cf8f9f895-kkj5r       1/1     Running             0
 #### Create the root issuer
 
-```console
+```{"stage":"cert-manager", "multiline_EOF":true, "runtime":"bash", "label":"create root issuer"}
````
nice default, only adding multiline_EOF as needed.

lavocatt replied:
I think this would be better named `"HereTag":"EOF"` since that's the name of that kind of tags.

gtully commented on docs/tutorials/send_receive_ingress_pem.md:
````diff
@@ -101,15 +110,14 @@ spec:
-send-receive-root-issuer   True             45s
+```{"stage":"cert-manager"}
+$ kubectl wait issuer --all --for=condition=Ready --namespace=send-receive-project
+issuer.cert-manager.io/send-receive-root-issuer condition met
````
even doing this automated testing process has caused us to improve the tutorial to use more dynamic commands and expose the conditions. great!

lavocatt replied:
I think this dynamic command can be improved, it's waiting for every issuer to be up, and we could make it so it only waits for the one we've just created.

gtully commented on docs/tutorials/send_receive_ingress_pem.md:
````diff
@@ -275,19 +281,37 @@ EOF
+Get the url for the ingress
 
+```{"stage":"etc", "variable":"INGRESS_URL", "runtime":"bash", "label":"get the ingress host"}
+$ kubectl get ingress send-receive-sslacceptor-0-svc-ing -o json | jq -r '.spec.rules[] | .host'
````
The INGRESS_URL, this seems a little odd... should that variable not be part of the md command?

$ export INGRESS_URL=$(get ingress send....)

same or CLUSTER_IP?

gtully replied:
for when manually working through the tutorial

gtully reviewed (commented):
This is fantastic, great work.

I guess pulling this command inline into a VerifyTutorialsTest for go test wold be trivial.

only comment of note is the use of env vars, to make sure those are visible to the non automated user.

lavocatt commented:
> This is fantastic, great work.
> 
> I guess pulling this command inline into a VerifyTutorialsTest for go test wold be trivial.
> 
> only comment of note is the use of env vars, to make sure those are visible to the non automated user.

Thanks a lot for your feedback! I will work on a better way for variables extraction as you requested.

lavocatt committed:
[#894] tutorial tester

Usage (to be executed at the root of the operator)

```
Usage of /tmp/go-build2452366115/b001/exe/tester:
  -dry-run
    	just list what would be executed without doing it
  -ignore-breakpoints
    	ignore the breakpoints
  -interactive
    	prompt to press enter between each chunk
  -list
    	just list the tutorials found
  -no-styling
    	disable spiners in cli
  -quiet
    	disable output
  -start-from string
    	start from a specific stage name
  -timeout int
    	the timeout in minutes for every executed command (default 10)
  -tutorial string
    	Run only a specific tutorial
  -tutorials-root string
    	where to find the tutorials (default "./docs/tutorials")
  -update-tutorials
    	update the chunk output section in the tutorials
  -verbose
    	print more logs
```

Checkout docs/tutorials/writing_tutorials.md for more details about how
to write executable tutorials.

brusdev commented on docs/tutorials/writing_tutorials.md:
````diff
@@ -0,0 +1,307 @@
+Let's have a verbose view:
+
+```
+╼> go run test/utils/tutorials/tester.go --verbose --run-only writing_tutorials.md
````
The option `--run-only` seems more a way to test tutorials than a way to select a tutorial, the test goal of the maven surefire plugin has a parameter with the name `test` to define the test to run, see https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html

lavocatt replied:
Yes I need a way to execute a single test at a time. I couldn't find a way to have positional arguments with the golang flag package easily, that's why I ended up using this option. You recommend to rename the `run-only` to `test`?

brusdev replied:
Also `tutorial` could be a good candidate, WDYT?

lavocatt replied:
I think that's way better than `run-only` thanks!

brusdev commented on docs/tutorials/writing_tutorials.md:
`````diff
@@ -0,0 +1,307 @@
+
+````
+```{"stage":"test4", "runtime":"bash", "HereTag":"EOF"}
+$ cat << EOF
`````
Could the tutorial tester autodetect command lines that ends with `<< EOF`?

brusdev replied:
I see your point but I suppose that that tutorial test could restrict the supported here tags to `<< EOF` simpolifing the usage without limiting too much the feature itself, unless you are aware of use case that could require another here tag.

brusdev reviewed (commented):
Great feature!!! It will improve the quality of the tutorials.

lavocatt commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -16,33 +16,97 @@ Before you start, you need to have access to a running Kubernetes cluster
+#### Start minikube with a parametrized dns domain name
+
+```{"stage":"init", "id":"minikube_start"}
+$ minikube start --dns-domain='demo.artemiscloud.io'
````
The default value of the cluster domain is  `cluster.local` see https://minikube.sigs.k8s.io/docs/commands/start/ which would mean that our ingresses url would end with `cluster.local` depending on the resolution method the user is chosing (NetworkManager's wildcard for instance) that would mean that the domain name wouldn't resolve correctly as `.local` is a reserved tld.

If you remember we had a similar discussion about 2 month ago on slack and we decided to use `demo.artemis.cloud` for the domain of the cluster to solve this issue.

brusdev replied:
I read `The cluster dns domain name used in the Kubernetes cluster` from https://minikube.sigs.k8s.io/docs/commands/start/. Why does the tutorial needs to change the dns domain name used in the Kubernetes cluster when the ingresse resources are used from the clients outside the Kubernetes cluster? Could you point me to the tutorial step that will fail without this option?

lavocatt replied:
I think I got confused. It's when we are using this https://github.com/artemiscloud/activemq-artemis-operator/blob/main/docs/help/hostname_resolution.md#minikubes-ingress-dns-plugin that we need to have it set..

lavocatt replied:
This has been resolved now. Thanks for chatting about the subject on slack!

lavocatt commented on docs/tutorials/writing_tutorials.md:
```diff
@@ -0,0 +1,307 @@
+The script itself is prefixed with `set -euo pipefail`, meaning that it will
+error at the first failing commands.
+
+##### `"HereTag":"EOF"`
```
good catch!

lavocatt commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -16,33 +16,97 @@ Before you start, you need to have access to a running Kubernetes cluster
-```
 
+```{"stage":"init", "runtime":"bash", "label":"wait for the operator to be running"}
+$ kubectl wait pod --for=condition=Ready --namespace=send-receive-project $(kubectl get pods --namespace send-receive-project | awk '/activemq-artemis-controller-manager/ {print $1;exit}')
````
That's shorter and simpler to read, let me try this out

lavocatt replied:
I think that in the end, it might just make sense to wait for all the pods of the send-receive-project namespace, since we've just created it, there's no risk of having to wait to other resources.

lavocatt replied:
I've done just that, is that good for you @brusdev ?

lavocatt commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -16,33 +16,97 @@ Before you start, you need to have access to a running Kubernetes cluster
+#### Start minikube with a parametrized dns domain name
+
+```{"stage":"init", "id":"minikube_start"}
+$ minikube start --dns-domain='demo.artemiscloud.io'
````
That's very neat! Then we can also delete with a profile too https://minikube.sigs.k8s.io/docs/commands/delete/
Then we would just have to make sure every tutorial has a unique profile.
Thanks!

brusdev replied:
Also a common profile for all tutorials could be fine if we need to reduce the startup time?

lavocatt commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -96,41 +195,47 @@ spec:
 
-For instance, you can add a new entry in your `/etc/hosts` to make this domain
-point to the associated address, in our example: `192.168.39.57`
+```{"stage":"etc", "env":["INGRESS_URL", "CLUSTER_IP"], "runtime":"bash", "label":"populate /etc/hosts", "id":"hosts_addition"}
+$ sudo echo "$CLUSTER_IP  $INGRESS_URL" | sudo tee -a /etc/hosts
````
That looks super useful.

Quick thought, to use it, that would mean that the `dns-name` of the cluster should be `${CLUSTER_IP}.nip.io`. But since we can't know the ip address of the cluster before starting it (even though that's often `192.168.1.2`) aren't stuck with a chicken and egg problem?

Unless I misunderstood the `dns-domain` of the minikube instance command line.

lavocatt replied:
We are using nip.io now, thanks for the suggestion

lavocatt commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -67,24 +131,59 @@ to configure the certificates.
-$ keytool -genkeypair -alias artemis -keyalg RSA -keysize 2048 -storetype PKCS12 -keystore broker.ks -validity 3000
-$ keytool -export -alias artemis -file broker.cert -keystore broker.ks
-$ keytool -import -v -trustcacerts -alias artemis -file broker.cert -keystore client.ts
+```{"stage":"cert-creation", "rootdir":"$tmpdir.1", "runtime":"bash", "label":"generate cert"}
+$ printf '000000\n000000\n\n\n\n\n\n\nyes\n' | keytool -genkeypair -alias artemis -keyalg RSA -keysize 2048 -storetype PKCS12 -keystore broker.ks -validity 3000
+$ printf '000000\n' | keytool -export -alias artemis -file broker.cert -keystore broker.ks
+$ printf '000000\n000000\nyes\n' | keytool -import -v -trustcacerts -alias artemis -file broker.cert -keystore client.ts
+
+Owner: CN=Unknown, OU=Unknown, O=Unknown, L=Unknown, ST=Unknown, C=Unknown
+Issuer: CN=Unknown, OU=Unknown, O=Unknown, L=Unknown, ST=Unknown, C=Unknown
+Serial number: 37372d20
+Valid from: Thu Aug 01 17:55:46 CEST 2024 until: Mon Oct 18 17:55:46 CEST 2032
+Certificate fingerprints:
+	 SHA1: 02:48:8B:E2:3A:6A:3D:64:F0:C8:F3:FA:61:FA:B3:55:59:AA:90:DE
+	 SHA256: E7:3F:0F:65:1F:E9:1F:39:F0:D3:0C:B5:0C:71:7A:E3:12:90:2A:AD:9B:77:26:3D:E8:0C:C8:DA:88:FE:CD:F2
+Signature algorithm name: SHA256withRSA
+Subject Public Key Algorithm: 2048-bit RSA key
+Version: 3
+
+Extensions: 
+
+#1: ObjectId: 2.5.29.14 Criticality=false
+SubjectKeyIdentifier [
+KeyIdentifier [
+0000: 15 45 6D 5C C1 0B AA A1   9F A0 DB 97 BD 82 5A 39  .Em\..........Z9
+0010: 66 04 D1 8E                                        f...
+]
+]
+
+```
````
What would be the common name to use in that case? Something defined by the host that is running the tutorial?

brusdev replied:
The ActiveMQArtemis CR supports the fields `ingressDomain` and `ingressHost` to control the ingress host. This means that the CN or the required subject alternative names are predicatble.

lavocatt replied:
I did experiment with that but I was unable to make it work unfortunately

brusdev commented on test/utils/tutorials/tester.go:
```diff
@@ -0,0 +1,621 @@
+	}
+	if chunk.Env != nil {
+		for _, v := range chunk.Env {
+			command.cmd.Env = append(command.cmd.Env, v+"="+variables[v])
```
When the field `Env` is set the os env is skipped, see https://pkg.go.dev/os/exec. To automatically restore preserve the previous command env and avoid env/varable declarations the tutorial tester could store them in a file at the end of a command (i.e `env > cmd.env`) and restore them at the beginning of a command (i.e. `set -a && source cmd.env && set +a`)

brusdev commented on test/utils/tutorials/tester.go:
```diff
@@ -0,0 +1,621 @@
+			isInChunk = false
+			searchForEof = false
+		} else if isInChunk {
+			command, isCommand := strings.CutPrefix(scanner.Text(), "$")
```
In this way the tutorial test doesn't need to search for the here tags or $

lavocatt replied:
Yes indeed, I've implemented this, and this is way better!

brusdev commented on test/utils/tutorials/tester.go:
````diff
@@ -0,0 +1,621 @@
+	var lineCounter = 0
+	for scanner.Scan() {
+		lineCounter += 1
+		if strings.HasPrefix(scanner.Text(), "```{") {
````
JSON is a good language for m2m protocols but for humans could be simpler a command line syntax, i.e.
```
```shell exec --stage="deply" --label=="minikube" --expected-content="created" --expected-result=0
```

The command line parameters could be parsed by using https://pkg.go.dev/flag

lavocatt replied:
As we've discussed outside the PR, there are other examples of people using JSON (https://github.com/artemiscloud/activemq-artemis-operator/pull/986).

I'm not against using something else, as long as the text object we get can be parsed and validated against some kind of schema. I don't have another idea than JSON for now.

Feel free to reopen if you get some inspiration :slightly_smiling_face:

gtully commented on docs/tutorials/writing_tutorials.md:
```diff
@@ -0,0 +1,315 @@
+In an executable tutorial, every line gets executed. Chunks that are executed are
+chunks that provide a valid json object with at least a stage configured in it.
+
+The code fend to create an executable chunk needs to have at least 3 back quotes
```
code fence in place of fend typo (I think)

lavocatt replied:
ah yes! I have to fix that

gtully reviewed (approved):
great work. very comprehensive

brusdev commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -148,64 +262,114 @@ a couple of parameters set:
+delete the minikube cluster and clean the `/etc/hosts` file.
+
+```{"stage":"teardown", "requires":"init/minikube_start"}
+minikube delete
````
```suggestion
minikube delete --profile tutorialtester
```

brusdev commented on docs/tutorials/send_receive_ingress_pem.md:
```diff
@@ -264,30 +356,36 @@ spec:
     - addressConfigurations."APP.JOBS".routingTypes=ANYCAST
     - addressConfigurations."APP.JOBS".queueConfigs."APP.JOBS".routingType=ANYCAST
     - addressConfigurations."APP.COMMANDS".routingTypes=MULTICAST
 EOF
```
Defining the variable INGRESS_HOST you don't need to split the writing of the file deploy.yml
```
INGRESS_HOST='ing.$(ITEM_NAME).$(CR_NAME)-$(BROKER_ORDINAL).$(CR_NAMESPACE).$(INGRESS_DOMAIN)'
kubectl apply -f - << EOF
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata:
  name: send-receive
  namespace: send-receive-project
spec:
  ingressDomain: demo.artemiscloud.io
  ingressDomain: ${CLUSTER_IP}.nip.io
  acceptors:
    - name: sslacceptor
      port: 62626
      sslEnabled: true
      expose: true
      exposeMode: ingress
      sslSecret: send-receive-sslacceptor-0-svc-ing-ptls
      ingressHost: ${INGRESS_HOST}
  resourceTemplates:
    - selector:
        kind: Ingress
@@ -264,30 +356,36 @@ spec:
        spec:
          tls:
            - hosts:
                - ing.sslacceptor.send-receive-0.send-receive-project.demo.artemiscloud.io
                - ing.sslacceptor.send-receive-0.send-receive-project.${CLUSTER_IP}.nip.io
              secretName: send-receive-sslacceptor-0-svc-ing-ptls
  brokerProperties:
    - addressConfigurations."APP.JOBS".routingTypes=ANYCAST
    - addressConfigurations."APP.JOBS".queueConfigs."APP.JOBS".routingType=ANYCAST
    - addressConfigurations."APP.COMMANDS".routingTypes=MULTICAST
EOF
```

brusdev commented on docs/tutorials/send_receive_ingress_pem.md:
````diff
@@ -51,24 +64,51 @@ If you are running your OpenShift cluster locally, you might not be able to
-$ kubectl get pod --namespace send-receive-project --watch
+
+```{"stage":"init"}
+kubectl wait pod --all --for=condition=Ready --namespace=send-receive-project --timeout=240s
````
The provided tls secret is created after 2~3 minutes and on a laptop with a slow internet connection the download of the container images could require other 2~3 minutes.
```suggestion
kubectl wait pod --all --for=condition=Ready --namespace=send-receive-project --timeout=600s
```

brusdev commented on docs/tutorials/send_receive_ingress.md:
````diff
@@ -67,24 +143,66 @@ to configure the certificates.
+```
+
+```{"stage":"cert-creation", "rootdir":"$tmpdir.1", "runtime":"bash", "label":"generate cert"}
+printf "000000\n000000\n${CLUSTER_IP}.nip.io\nArtemisCloud\nRed Hat\nGrenoble\nAuvergne Rhône Alpes\nFR\nyes\n" | keytool -genkeypair -alias artemis -keyalg RSA -keysize 2048 -storetype PKCS12 -keystore broker.ks -validity 3000
+printf '000000\n' | keytool -export -alias artemis -file broker.cert -keystore broker.ks
+printf '000000\n000000\nyes\n' | keytool -import -v -trustcacerts -alias artemis -file broker.cert -keystore client.ts
````
That could be food for a future PR
